crates/reth/statediff/src/lib.rs

```diff
@@ -0,0 +1,316 @@
+#![cfg_attr(not(test), warn(unused_crate_dependencies))]
+
+use alloy_primitives::U256;
+use revm::db::{BundleAccount, BundleState};
+use revm_primitives::{AccountInfo, Address, Bytecode, HashMap, B256};
+use serde::{Deserialize, Serialize};
+
+/// Represents full state diff for the block together with original values.
+#[derive(Clone, Default, Debug, Deserialize, Serialize, PartialEq)]
+pub struct BlockStateDiff {
+    /// Account state.
+    pub state: HashMap<Address, BundleAccount>,
+    /// All created contracts in this block.
+    pub contracts: HashMap<B256, Bytecode>,
+}
+
+impl From<BundleState> for BlockStateDiff {
+    fn from(value: BundleState) -> Self {
+        Self {
+            state: value.state,
+            contracts: value.contracts,
+        }
+    }
+}
+
+/// A representation of several [`BlockStateDiff`] ready to be serialized and posted on DA.
+///
+/// N.B. "original" counterparts are needed to correctly construct the [`BatchStateDiff`]
+/// for the range of blocks and not include changes for the keys whose values
+/// were changed inside batch, but end up having the same value as before the batch.
```

Comment:
This seems like it's not just a diff then?  We should have some builder type that remembers the original values and tracks modifications we're making, comparing against the originals, and updates an inner diff structure we extract at the end?

Reply:
Well, agree. Introduced a thin builder type that does pretty much the same.